File: 5. Longest Palindromic Substring/Soluction.c
#include <stdio.h>
#include <stdlib.h>

#define TAM_ARRAY 10

char *longestPalindromicSubstring(char array[]){
    char *newArray = (char *) malloc (sizeof(char));
    int tamNewArray = 0;
    int index[2] = {0,0};

    for (int i = 0; i < TAM_ARRAY-1; i++){
        for (int j = i+1; j < TAM_ARRAY-1; j++){
            if (array[i] == array[j]){
                int tamNewArrayTemp = 0;
                for (int k = 0; k < j-i; k++){
                    if(array[i+k] == array[j-k]){
                        tamNewArrayTemp = k;

                    }else{
                        tamNewArrayTemp = 0;
                        k = j-i+1;
                    }
                    
                    if (k == j-i-1 && tamNewArray < j-i+1){
                        index[0] = i;
                        index[1] = j;
                        tamNewArray = j-i+1;
                    }
                }
            }
        }
    }

    newArray = (char *) realloc (newArray, index[1]-index[0]+1 * sizeof(char));
    for (int i = index[0], j = 0; j <= index[1]; i++,j++)
        newArray[j] = array[i];
    newArray[index[1]-index[0]+1] = '\0';
          
    return newArray;
}

int main(){
    char array[TAM_ARRAY] = "souluo";
    char *arrayReduce;
    arrayReduce = longestPalindromicSubstring(array);
    
    printf("[%s, %s]", array, arrayReduce);
    
    free(arrayReduce);
    return 0;
}

File: 7. Reverse Integer/Soluction.java
import java.util.Scanner;

public class Soluction {

    public static class Resultado {
        private int[] indices;

        public Resultado() {
            this.indices = new int[2];
        }

        public int[] getIndices() {
            return indices;
        }

        public void setIndices(int i, int j) {
            this.indices[0] = i;
            this.indices[1] = j;
        }

        public boolean isEncontrado() {
            return indices[0] != 0 || indices[1] != 0;
        }
    }

    public static Resultado buscarIndices(int[] array, int target) {
        int TAM_ARRAY = array.length;
        Resultado resultado = new Resultado();

        for (int i = 0; i < TAM_ARRAY - 1; i++) {
            for (int j = i + 1; j < TAM_ARRAY; j++) {
                if (array[i] + array[j] == target) {
                    resultado.setIndices(i, j);
                    return resultado;
                }
            }
        }

        resultado.setIndices(0, 0);
        return resultado;
    }

    public static void main(String[] args) {
        int[] array = {-1, 2, 3, 4, 5, 6};
        Scanner scanner = new Scanner(System.in);

        System.out.println("Insira um valor que você busca encontrar a soma: ");
        int target = scanner.nextInt();

        Resultado resultado = buscarIndices(array, target);

        if (resultado.isEncontrado()) {
            System.out.printf("[%d, %d]%n", resultado.getIndices()[0], resultado.getIndices()[1]);
        } else {
            System.out.println("Não temos valores que resultam nessa soma");
        }

        scanner.close();
    }
}

// #define TAM_ARRAY 6

// char *longestSubstring(char array[]){
//     char *newArray = (char *) malloc (TAM_ARRAY * sizeof(char));
//     int positionNewArray = 0;
//     int ahLetra = 0;

//     for (int i = 0; i < TAM_ARRAY-1; i++){
//         for (int j = 0; j < positionNewArray; j++){
//             if(array[i] == newArray[j])
//                 ahLetra = 1;
//         }
//         if (!ahLetra){
//             newArray[positionNewArray] = array[i];
//             positionNewArray++;
//         }
//         ahLetra = 0;
//     }
//     newArray[positionNewArray] = '\0';
          
//     return newArray;
// }

// int main(){
//     char array[TAM_ARRAY] = "abbb";
//     char *arrayReduce;

//     arrayReduce = longestSubstring(array);
    
//     printf("[%s, %s]", array, arrayReduce);
    
//     free(arrayReduce);

//     return 0;
// }